Mathlib/LinearAlgebra/FreeModule/Finite/Matrix.lean

```diff
@@ -21,77 +21,93 @@ We provide some instances for finite and free modules involving matrices.
 -/
 
 
-universe u v w
+universe u u' v w
 
-variable (R : Type u) (M : Type v) (N : Type w)
+variable (R : Type u) (S : Type u') (M : Type v) (N : Type w)
 
-open Module.Free (chooseBasis)
+open Module.Free (chooseBasis ChooseBasisIndex)
 
 open FiniteDimensional (finrank)
 
-section CommRing
+section Ring
 
-variable [CommRing R] [AddCommGroup M] [Module R M] [Module.Free R M]
+variable [Ring R] [Ring S] [AddCommGroup M] [Module R M] [Module.Free R M] [Module.Finite R M]
+variable [AddCommGroup N] [Module R N] [Module S N] [SMulCommClass R S N]
 
-variable [AddCommGroup N] [Module R N] [Module.Free R N]
+private noncomputable def linearMapEquivFun : (M →ₗ[R] N) ≃ₗ[S] ChooseBasisIndex R M → N :=
+  (chooseBasis R M).repr.congrLeft N S ≪≫ₗ (Finsupp.lsum S).symm ≪≫ₗ
+    LinearEquiv.piCongrRight fun _ ↦ LinearMap.ringLmapEquivSelf R S N
 
-instance Module.Free.linearMap [Module.Finite R M] [Module.Finite R N] :
-    Module.Free R (M →ₗ[R] N) := by
-  cases subsingleton_or_nontrivial R
-  · apply Module.Free.of_subsingleton'
-  classical exact
-      Module.Free.of_equiv (LinearMap.toMatrix (chooseBasis R M) (chooseBasis R N)).symm
+instance Module.Free.linearMap [Module.Free S N] : Module.Free S (M →ₗ[R] N) :=
+  Module.Free.of_equiv (linearMapEquivFun R S M N).symm
 #align module.free.linear_map Module.Free.linearMap
 
-variable {R}
-
-instance Module.Finite.linearMap [Module.Finite R M] [Module.Finite R N] :
-    Module.Finite R (M →ₗ[R] N) := by
-  cases subsingleton_or_nontrivial R
-  · infer_instance
-  classical
-    have f := (LinearMap.toMatrix (chooseBasis R M) (chooseBasis R N)).symm
-    exact Module.Finite.of_surjective f.toLinearMap (LinearEquiv.surjective f)
+instance Module.Finite.linearMap [Module.Finite S N] : Module.Finite S (M →ₗ[R] N) :=
+  Module.Finite.equiv (linearMapEquivFun R S M N).symm
 #align module.finite.linear_map Module.Finite.linearMap
 
-end CommRing
+variable [StrongRankCondition R] [StrongRankCondition S] [Module.Free S N]
 
-section Integer
+open Cardinal
+theorem FiniteDimensional.rank_linearMap :
```

Comment:
Can you add a primed simpler version without `lift` when things are in the same universe?

Reply:
I'm not sure that would be useful, because the LHS doesn't have `lift` in it, so you can rewrite in the forward direction without problem and then use `lift_id` to simplify the RHS when M and N are in the same universe. I originally [introduced](https://github.com/leanprover-community/mathlib4/pull/9677/commits/cc18fb2eda3061f0a1d27e4ad4af1372eae8f17d#diff-6cfabfe01c4ebd977e118300cd5031161533848345f9e5afad4f97795368ebcaR36) a variable M' for the version that you suggest, but later decided to remove it.

Reply:
I see, good point.

Reply:
Thanks!
bors r+